connection_manager.go

```diff
@@ -0,0 +1,393 @@
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// openConn opens a connection to the given cluster
+func (cm *connectionManager) openConn(cl Cluster) (Connection, error) {
+	d := cm.drivers[cl.Type().Name()]
+	if d == nil {
+		cm.log.Errorf("unsupported driver type %s for %s:%s",
+			cl.Type().Name(), cl.Database(), cl.Name())
+		return nil, errStorageTypeUnsupported
+	}
+
+	cfg := proto.Clone(d.ConfigType())
+	if err := cl.Config().Unmarshal(cfg); err != nil {
+		cm.log.Errorf("failed to unmarshal config v%d for %s:%s: %v",
+			cl.Config().Version(), cl.Database(), cl.Name(), err)
+		return nil, err
+	}
+
+	return d.OpenConnection(cfg)
+}
+
+// registerConn registers a newly opened connection to the given database cluster
+func (cm *connectionManager) registerConn(cl Cluster, cw ClusterWatch, conn Connection) error {
+	key := fmtConnKeyFromCluster(cl)
+
+	cm.Lock()
+	defer cm.Unlock()
+
+	if cm.closed {
+		cw.Close()
+		conn.Close()
+		return errClosed
+	}
+
+	// Register the watch and connection
+	cm.conns[key], cm.watches[key] = conn, cw
+
+	// Spin up a goroutine to monitor the cluster for updates
+	go cm.watchCluster(cl, cw, conn)
+	cm.wg.Add(1)
+
+	return nil
+}
+
+// watchCluster monitors the given cluster for config updates
+func (cm *connectionManager) watchCluster(cl Cluster, cw ClusterWatch, conn Connection) {
+	cm.log.Infof("watching %s:%s", cl.Database(), cl.Name())
+	for range cw.C() {
+		cm.reconfigureConn(cw.Get(), conn)
+	}
+
+	cm.cleanupConn(cl.Database(), cl.Name())
+	cm.wg.Done()
+}
+
+// reconfigureConn reconfigures a connection with new cluster configuration
+func (cm *connectionManager) reconfigureConn(cl Cluster, existing Connection) error {
+	key := fmtConnKeyFromCluster(cl)
+	d := cm.drivers[cl.Type().Name()]
+	if d == nil {
+		cm.log.Errorf("received update v%d for %s:%s: unsupported storage type %s",
+			cl.Config().Version(), cl.Database(), cl.Name(), cl.Type().Name())
+		return errStorageTypeUnsupported
+	}
+
+	cm.log.Infof("received update v%d for %s:%s", cl.Config().Version(), cl.Database(), cl.Name())
+
+	cfg := proto.Clone(d.ConfigType())
+	if err := cl.Config().Unmarshal(cfg); err != nil {
+		// Don't bail out, just log and metric
+		cm.log.Errorf("invalid config v%d for %s:%s: %v",
+			cl.Config().Version(), cl.Database(), cl.Name(), err)
+		return err
+	}
+
+	newConn, err := d.ReconfigureConnection(existing, cfg)
+	if err != nil {
+		cm.log.Errorf("could not reconfigure conn for %s:%d v%d: %v",
+			cl.Database(), cl.Name(), cl.Config().Version(), err)
+		return err
+	}
+
+	cm.Lock()
+	cm.conns[key] = newConn
+	cm.Unlock()
+	return nil
+}
+
+// cleanupConn cleans up the connection to the given database and cluster
+func (cm *connectionManager) cleanupConn(database, cluster string) {
+	cm.log.Infof("cleaning up connection to %s:%s", database, cluster)
+
+	key := fmtConnKey(database, cluster)
+
+	cm.Lock()
+	conn := cm.conns[key]
+	delete(cm.conns, key)
+	delete(cm.watches, key)
```

Comment:
No, the watch is closed in `connectionManager.Close()`.  Closing the watch is what causes the background goroutine to break out of its watch loop and perform this cleanup code.